The PodOverhead feature is GA (#32328)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/pod-overhead.md

<!-- overview -->

{{< feature-state for_k8s_version="v1.24" state="stable" >}}

When you run a Pod on a Node, the Pod itself takes an amount of system resources. These
resources are additional to the resources needed to run the container(s) inside the Pod.

> In Kubernetes, "Pod Overhead" is a way for to account for the resources consumed by the Pod

Reviewer: We [use italics to introduce new concepts](https://kubernetes.io/docs/contribute/style/style-guide/#use-italics-to-define-or-introduce-new-terms)
```suggestion
In Kubernetes, _Pod overhead_ is a way for to account for the resources consumed by the Pod
```

<!-- overview -->

{{< feature-state for_k8s_version="v1.24" state="stable" >}}

When you run a Pod on a Node, the Pod itself takes an amount of system resources. These
resources are additional to the resources needed to run the container(s) inside the Pod.

> 'Pod Overhead' is a feature for accounting for the resources consumed by the Pod infrastructure

Reviewer: We should document behaviors over features.
```suggestion
In Kubernetes,  _pod overhead_ is a way to account for the resources consumed by the Pod infrastructure
```

on top of the container requests & limits.

<!-- body -->

In Kubernetes, the Pod's overhead is set at
[admission](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks)
time according to the overhead associated with the Pod's
[RuntimeClass](/docs/concepts/containers/runtime-class/).

A pod's overhead is considered in addition to the sum of container resource requests when
scheduling a Pod. Similarly, the kubelet will include the Pod overhead when sizing the Pod cgroup,
and when carrying out Pod eviction ranking.

> ## Enabling Pod Overhead {#set-up}

Reviewer: ```suggestion
## Configuring pod overhead {#set-up}
```
or, actually, omit the header. The text at the start of the next section makes much the same point.

You need to make sure a `RuntimeClass` is utilized which defines the `overhead` field.

## Usage example

> To use the Pod overhead feature, you need a RuntimeClass that defines the `overhead` field. As

Reviewer: ```suggestion
To work with Pod overhead, you need a RuntimeClass that defines the `overhead` field. As
```

the Pod will be rejected. In the given example, since only the RuntimeClass name is specified, the admission controller mutates the Pod
to include an `overhead`.

> After the RuntimeClass admission controller, you can check the updated Pod spec:

Reviewer: ```suggestion
After the RuntimeClass admission controller has made modifications, you can check the updated
overhead value:
```

(I wonder, could we also show this with `kubectl ... --dry-run=server`?)

Author: I'm not sure if dry run works for this case.

requests plus the `overhead` defined in the PodSpec.

Looking at our example, verify the container requests for the workload:

```bash
kubectl get pod test-pod -o jsonpath='{.spec.containers[*].resources.limits}'
```

The total container requests are 2000m CPU and 200MiB of memory:

```
map[cpu: 500m memory:100Mi] map[cpu:1500m memory:100Mi]
```

Check this against what is observed by the node:

```bash
kubectl describe node | grep test-pod -B2
```

> The output shows 2250m CPU and 320MiB of memory are requested, which includes Pod overhead:

Reviewer: How about:
```suggestion
The output shows requests for 2.250 CPU, and for 320MiB of memory. These requests includes Pod overhead:
```

I think “2.25” is more meaningful to people who aren't familiar with Kubernetes quantities or with Docker CPU fractions.

Author: I don't think so. The `2250m` expression is the new norm in Kubernetes. People have to get used to that.

sudo crictl inspectp -o=json $POD_ID | grep cgroupsPath
```

The resulting cgroup path includes the Pod's `pause` container. The Pod level cgroup is one directory above.

```
  "cgroupsPath": "/kubepods/podd7f4b509-cf94-4951-9417-d1087c92a5b2/7ccf55aee35dd16aca4189c952d83487297f3cd760f1bbf09620e206e7d0c27a"
```

In this specific case, the pod cgroup path is `kubepods/podd7f4b509-cf94-4951-9417-d1087c92a5b2`.
Verify the Pod level cgroup setting for memory:

```bash
# Run this on the node where the Pod is scheduled.
# Also, change the name of the cgroup to match the cgroup allocated for your pod.
 cat /sys/fs/cgroup/memory/kubepods/podd7f4b509-cf94-4951-9417-d1087c92a5b2/memory.limit_in_bytes
```

This is 320 MiB, as expected:

```
335544320
```

### Observability

A `kube_pod_overhead` metric is available in [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics)
to help identify when Pod overhead is being utilized and to help observe stability of workloads
running with a defined overhead. This functionality is not available in the 1.9 release of
`kube-state-metrics`, but is expected in a following release. Users will need to build

> `kube-state-metrics` from source in the meantime.

Reviewer: > This functionality is not available in the 1.9 release of `kube-state-metrics`, but is expected in a following release. Users will need to build `kube-state-metrics` from source in the meantime.

We [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future); I'd drop this

Author: Double checked  the project and the metrics. The text needs some update.

## {{% heading "whatsnext" %}}

* [RuntimeClass](/docs/concepts/containers/runtime-class/)

> * [PodOverhead Design](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/688-pod-overhead)

Reviewer: ```suggestion
* Read the [PodOverhead](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/688-pod-overhead) enhancement proposal for extra context
```
